bee-protocol/src/worker/confirmer.rs

```diff
@@ -0,0 +1,202 @@
+    }
+}
+
+async fn process<B: StorageBackend>(tangle: &MsTangle<B>, milestone: Milestone, index: MilestoneIndex) {
+    let message_id = milestone.message_id();
+
+    if let Some((parent1, parent2)) = tangle
+        .get(message_id)
+        .await
+        .map(|message| (*message.parent1(), *message.parent2()))
+    {
+        // Confirm the past cone of this milestone, and return all newly confir
+        let confirmed = confirm_past_cone(tangle, parent1, parent2, index).await;
+
+        // Propagate new confirmation states
+        update_otrsi_ytrsi(tangle, confirmed).await;
+    }
+
+    // Update tip pool after all values got updated.
+    tangle.update_tip_scores().await;
+}
+
+async fn confirm_past_cone<B: StorageBackend>(
+    tangle: &MsTangle<B>,
+    parent1: MessageId,
+    parent2: MessageId,
+    index: MilestoneIndex,
+) -> HashSet<MessageId> {
+    let mut parents = vec![parent1, parent2];
+    let mut confirmed = HashSet::new();
+
+    while let Some(parent_id) = parents.pop() {
+        // Our stop conditions. Note that the order of calls is important (from cheap to more expensive) for performance
+        // reasons.
+        if confirmed.contains(&parent_id)
+            || tangle.is_solid_entry_point(&parent_id)
+            || tangle.get_metadata(&parent_id).await.unwrap().flags().is_confirmed()
+        {
+            continue;
+        }
+
+        tangle
+            .update_metadata(&parent_id, |metadata| {
+                // TODO: Throw one of those indexes away ;)
+                metadata.set_milestone_index(index);
+                metadata.set_cone_index(index);
+                metadata.set_otrsi(IndexId(index, parent_id));
+                metadata.set_ytrsi(IndexId(index, parent_id));
+                metadata.flags_mut().set_confirmed(true);
+            })
+            .await;
+
+        if let Some((grand_parent1, grand_parent2)) = tangle
+            .get(&parent_id)
+            .await
+            .map(|parent| (*parent.parent1(), *parent.parent2()))
+        {
+            parents.push(grand_parent1);
+            parents.push(grand_parent2);
+        }
+
+        // Preferably we would only collect the 'root messages/transactions'. They are defined as being confirmed by
+        // a milestone, but at least one of their children is not confirmed yet. One can think of them as an attachment
+        // point for new messages to the main tangle. It is ensured however, that this set *contains* the root messages
+        // as well, and during the future walk we will skip already confirmed children, which shouldn't be a performance
+        // issue.
+        confirmed.insert(parent_id);
+    }
+
+    debug!("Confirmed {} messages.", confirmed.len());
+
+    confirmed
+}
+
+// NOTE: so once a milestone comes in we have to walk the future cones of the root transactions and update their
+// OTRSI and YTRSI; during that time we need to block the propagator, otherwise it will propagate outdated data.
```

Comment:
"we need to block the propagator" is this implemented?

Reply:
nope. I am pretty sure, however, that not pausing it, leads to a data race.